test/distributed/_composable/fully_shard/test_fully_shard_mixed_precision.py

```diff
@@ -31,31 +31,160 @@
 class TestMixedPrecision(FSDPTest):
     """Tests ``fully_shard`` with mixed precision."""
 
+    EXPECTED_CAST_DTYPES = {
+        # (cast_root_forward_inputs_submodule, cast_forward_inputs_submodule, use_root_no_params)
+        (True, True, True): (
+            "cast_root_cast_child_no_root_params",
+            torch.float16,
+            torch.float16,
+            torch.float16,
+        ),
+        (True, True, False): (
+            "cast_root_cast_child_root_params",
+            torch.float32,
+            torch.float32,
+            torch.float16,
+        ),
+        (True, False, True): (
+            "cast_root_no_cast_child_no_root_params",
+            torch.float16,
+            torch.float16,
+            torch.float16,
+        ),
+        (True, False, False): (
+            "cast_root_no_cast_child_root_params",
+            torch.float32,
+            torch.float32,
+            torch.float16,
+        ),
+        (False, True, True): (
+            "no_cast_root_cast_child_no_root_params",
+            torch.float32,
+            torch.float16,
+            torch.float16,
+        ),
+        (False, True, False): (
+            "no_cast_root_cast_child_root_params",
+            torch.float32,
+            torch.float32,
+            torch.float16,
+        ),
+        (False, False, True): (
+            "no_cast_root_no_cast_child_no_root_params",
+            torch.float32,
+            torch.float32,
+            None,
+        ),
+        # (False, False, True): ('no_cast_root_no_cast_child_no_root_params',
+        #                        torch.float32, torch.float32, torch.float32),  # expected result after eval mode fix
+        (False, False, False): (
+            "no_cast_root_no_cast_child_root_params",
+            torch.float32,
+            torch.float32,
+            torch.float32,
+        ),
+    }
+
     @property
     def world_size(self):
         return 2
 
     @skip_if_lt_x_gpu(2)
-    def test_float16_on_one_submodule(self):
-        forward_inputs: Dict[nn.Module, torch.Tensor] = {}
-        float16 = MixedPrecision(param_dtype=torch.float16, cast_forward_inputs=True)
+    def test_float16_cast_forward(self):
+        self.run_subtests(
+            {
+                "cast_root_forward_inputs_submodule": [True, False],
+                "cast_forward_inputs_submodule": [True, False],
+                "use_root_no_params": [True, False],
+            },
+            self._test_float16_cast_forward,
+        )
+
+    def _test_float16_cast_forward(
+        self,
+        cast_root_forward_inputs_submodule: bool,
+        cast_forward_inputs_submodule: bool,
+        use_root_no_params: bool,
+    ):
+        self.cast_root_forward_inputs_submodule = cast_root_forward_inputs_submodule
+        self.cast_forward_inputs_submodule = cast_forward_inputs_submodule
+        x, fsdp = self._input_and_model_init(use_root_no_params=use_root_no_params)
+        # self._validate_eval(x, fsdp)
```

Comment:
one can uncomment this function once a broader solution to #99545 is decided upon